src/emulator/kernel/src/thread/sync_primitives.cpp

```diff
@@ -268,5 +333,134 @@ int signal_sema(KernelState &kernel, const char *export_name, SceUID semaid, int
+            export_name, weight == SyncWeight::Light ? "lwcond" : "cond", condvar->uid, condvar->name, condvar->attr, condvar->associated_mutex->uid);
+    }
+
+    const std::lock_guard<std::mutex> lock(condvar->mutex);
+
+    const ThreadStatePtr thread = lock_and_find(thread_id, kernel.threads, kernel.mutex);
+
+    bool is_fifo = (condvar->attr & SCE_KERNEL_ATTR_TH_FIFO);
+
+    if (auto error = mutex_unlock_impl(kernel, export_name, thread_id, 1, condvar->associated_mutex))
+        return error;
+
+    const std::lock_guard<std::mutex> lock2(thread->mutex);
+    assert(thread->to_do == ThreadToDo::run);
+    thread->to_do = ThreadToDo::wait;
+
+    WaitingThreadData data;
+    data.thread = thread;
+    data.priority_condvar = is_fifo ? 0 : thread->priority;
+
+    condvar->waiting_threads.emplace(data);
+    stop(*thread->cpu);
+
+    return SCE_KERNEL_OK;
+}
+
+int condvar_signal(KernelState &kernel, const char *export_name, SceUID thread_id, SceUID condid, Condvar::SignalTarget signal_target, SyncWeight weight) {
+    // TODO Don't lock twice.
+    const CondvarPtr condvar = lock_and_find(condid, kernel.condvars, kernel.mutex);
+    if (!condvar) {
+        return unknown_cond_id(export_name, weight);
+    }
+
+    if (LOG_SYNC_PRIMITIVES) {
+        LOG_DEBUG("{}: Signaling {}: uid:{} name:\"{}\" attr:{} assoc_mutexid:{}",
+            export_name, weight == SyncWeight::Light ? "lwcond" : "cond", condvar->uid, condvar->name, condvar->attr, condvar->associated_mutex->uid);
+    }
+
+    const std::lock_guard<std::mutex> lock(condvar->mutex);
+
+    const auto target_type = signal_target.type;
+
+    WaitingThreadData waiting_thread_data;
+    if (target_type == Condvar::SignalTarget::Type::Specific) {
+        ThreadStatePtr target_thread = lock_and_find(signal_target.thread_id, kernel.threads, kernel.mutex);
+
+        // Search for specified waiting thread
+        // TODO: this is shit, should implement our own priority queue that accomodates iteration
+        WaitingThreadQueue waiting_threads_copy = condvar->waiting_threads;
+
+        while (!waiting_threads_copy.empty()) {
+            const WaitingThreadData temp_data = waiting_threads_copy.top();
+            if (temp_data.thread == target_thread)
+                waiting_thread_data = std::move(temp_data);
+            waiting_threads_copy.pop();
+        }
+    }
+
+    while (condvar->waiting_threads.size() > 0) {
+        ThreadStatePtr waiting_thread;
+        if (target_type != Condvar::SignalTarget::Type::Specific) {
+            // Pick first waiting thread (based on priority)
+            waiting_thread_data = condvar->waiting_threads.top();
+        }
+        waiting_thread = waiting_thread_data.thread;
+
+        {
+            const std::lock_guard<std::mutex> lock2(waiting_thread->mutex);
+
+            assert(waiting_thread->to_do == ThreadToDo::wait);
+            waiting_thread->to_do = ThreadToDo::run;
+            condvar->waiting_threads.pop();
```

Comment:
you don't remove target_thread when specific

Reply:
Good catch 👍 .
I don't feel like implementing a priority queue now, I will make a couple of helper functions to deal with it until we do.